src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java

```diff
@@ -180,19 +185,29 @@ public BitbucketCommit call() throws Exception {
 
     private final String baseURL;
 
+    private final BitbucketServerWebhookImplementation webhookImplementation;
+
     @Deprecated
     public BitbucketServerAPIClient(@NonNull String baseURL, @NonNull String owner, @CheckForNull String repositoryName,
                                     @CheckForNull StandardUsernamePasswordCredentials credentials, boolean userCentric) {
-        this(baseURL, owner, repositoryName, credentials != null ? new BitbucketUsernamePasswordAuthenticator(credentials) : null, userCentric);
+        this(baseURL, owner, repositoryName, credentials != null ? new BitbucketUsernamePasswordAuthenticator(credentials) : null,
+            userCentric, BitbucketServerEndpoint.findWebhookImplementation(baseURL));
     }
 
     public BitbucketServerAPIClient(@NonNull String baseURL, @NonNull String owner, @CheckForNull String repositoryName,
                                     @CheckForNull BitbucketAuthenticator authenticator, boolean userCentric) {
+        this(baseURL, owner, repositoryName, authenticator, userCentric, BitbucketServerEndpoint.findWebhookImplementation(baseURL));
+    }
+
+    public BitbucketServerAPIClient(@NonNull String baseURL, @NonNull String owner, @CheckForNull String repositoryName,
+                                    @CheckForNull BitbucketAuthenticator authenticator, boolean userCentric,
+                                    @NonNull BitbucketServerWebhookImplementation webhookImplementation) {
         this.authenticator = authenticator;
         this.userCentric = userCentric;
         this.owner = owner;
         this.repositoryName = repositoryName;
         this.baseURL = Util.removeTrailingSlash(baseURL);
+        this.webhookImplementation = requireNonNull(webhookImplementation);
```

Comment:
This creates a null pointer for people upgrading to this version.

Reply:
Not sure how this can happen? Looking at the Stack Trace that @aairey kindly provided, I'm not sure how this can be null:

`at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.<init>(BitbucketServerAPIClient.java:199)`
https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/b12aea784da5cf14ba429c8f7e889e7ed9a6bcf2/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java#L197-L200

Here, `findWebhookImplementation` is called, which retrieves the webhook implementation from the endpoint:
https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/b12aea784da5cf14ba429c8f7e889e7ed9a6bcf2/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java#L104-L112

In the endpoint, `webhookImplementation` is initialized to `PLUGIN`.
https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/b12aea784da5cf14ba429c8f7e889e7ed9a6bcf2/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java#L79-L80

Then, there's a `@DataBoundSetter` for it, which also has requireNonNull in it:
https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/b12aea784da5cf14ba429c8f7e889e7ed9a6bcf2/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java#L159-L162

Am I missing something? I suspect this is some Jenkins deserialization magic? Would be nice if someone could provide a hint? Everything is annotated with `@NonNull`, and static code analysis is also happy.

Reply:
pretty sure the unit test can replicate 🤭

Reply:
It's Jenkins/XStream deserialization. Internally, sun.misc.Unsafe::allocateInstance is used to create new instances, which bypasses all constructor code, leaving webhookImpementation uninitialized, if it's not already present in the stored configuration.

I've a test case for that. Will figure out how to fix.

Reply:
Add a `readResolve` Method

Reply:
Thanks @stephenc, apparently, that did the trick!

Reply:
Fixed in e9b7774:

https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/e9b777496513dc5882528c06a8409fb7008ee66e/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java#L155-L161